cpp/ql/src/semmle/code/cpp/models/implementations/StdContainer.qll

```diff
@@ -88,6 +88,32 @@ class StdSequenceContainerFrontBack extends TaintFunction {
   }
 }
 
+/**
+ * The standard container function `assign`.
+ */
+class StdSequenceContainerAssign extends TaintFunction {
+  StdSequenceContainerAssign() {
+    this.hasQualifiedName("std", ["vector", "deque", "list", "forward_list"], "assign")
+  }
+
+  /**
+   * Gets the index of a parameter to this function that is a reference to the
+   * value type of the container.
+   */
+  int getAValueTypeParameterIndex() {
+    getParameter(result).getUnspecifiedType() = getDeclaringType().getTemplateArgument(0) // i.e. the `T` of this `std::vector<T>`
```

Comment:
Which overload is this for? On https://en.cppreference.com/w/cpp/container/vector/assign I don't see any overload that takes a `T`. Overload 1 takes a `const T&`, and the others take iterators and initializer lists.

Do you intend to handle the two other overloads in a follow-up PR? Why is it necessary to do such an elaborate detection to find the parameter if you're only targeting one overload?

Reply:
1. This was needed in practice to catch the taint on flow on vector.cpp line 222:
```
v2.assign(100, ns_int::source());
```
At the time I thought it was strange as it should be an `int &`.  Looking at it now I think it's actually calling `template<class InputIterator> void assign(InputIterator first, InputIterator last);` with `InputIterator = int` rather than `void assign(size_type n, const T& u);`.  I'm now quite confused about how this is ever expected to work when `T` is compatible with `size_type`!

2. Yes, I intend to handle the iterator overloads in a follow-up PR after Roberts iterator PR (https://github.com/github/codeql/pull/4102) is merged.  It may turn out that this simplifies the code (just excluding the `size_t` parameter).

Reply:
I think `v2.assign(100, ns_int::source());` correctly calls the first overload on https://en.cppreference.com/w/cpp/container/vector/assign. There's nothing wrong with doing this: https://ideone.com/urRYCZ

Reply:
The real-world C++ libraries apparently use SFINAE to remove the iterator overload for non-iterator types: https://github.com/gcc-mirror/gcc/blob/6b3034eaba83935d9f6dfb20d2efbdb34b5b00bf/libstdc%2B%2B-v3/include/bits/stl_vector.h#L764-L779.

So apparently we should be careful not to take the type signatures on cppreference.com literally.

Reply:
> So apparently we should be careful not to take the type signatures on cppreference.com literally.

Or apparently the ISO C++ Standards. :(

I'll see if I can figure out how to make the header behave properly.

Reply:
Fixed.  Obviously I had to vary the header from what the standard says a little bit to get this to work.